### PR TITLE
test: use graphql spec files for tests

### PR DESCRIPTION
**Pull request by sujeetsr**
- [x] Ability to execute multiple GraphQL queries within a file.
- [x] Ability to specify the client SDL, service SDL and multiple queries within a file.
- [x] Ability to specify multiple validation errors in the graphql file.
- [ ] Cleanup file structure

### CODE REVIEW

File: runtime/src/main/scala/tailcall/runtime/model/ConstantType.scala [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package tailcall.runtime.model
+
+import tailcall.runtime.DirectiveCodec
+import zio.json.JsonCodec
+import zio.schema.annotation.caseName
+import zio.schema.codec.JsonCodec.jsonCodec
+import zio.schema.{DeriveSchema, Schema}
+
+@caseName("constant")
+final case class ConstantType(value: String)
```
Comment (tusharmath) [inline runtime/src/main/scala/tailcall/runtime/model/ConstantType.scala:10]:
```suggestion
final case class ConstantType(value: Json)
```
Using Json will move the validation to the step where we read a config. You won't need to write code to validate if the string is Json.

File: runtime/src/test/resources/tailcall/runtime/graphql/execution/test-dictionary.graphql [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    a 
+    b {
+      key
+      value
+    }
+  }
+}
+
+#> query-response
+# {"z":{"a":1,"b":[{"key":"k1","value":1},{"key":"k2","value":2},{"key":"k3","value":3}]}}
```
Comment (tusharmath) [inline runtime/src/test/resources/tailcall/runtime/graphql/execution/test-dictionary.graphql:32]:
```suggestion
query @expect(output: {"z":{"a":1,"b":[{"key":"k1","value":1},{"key":"k2","value":2},{"key":"k3","value":3}]}}) {
  z {
    a 
    b {
      key
      value
    }
  }
}

```

This kind of syntax will allow us to write multiple queries in the same file and assert the result directly.

File: runtime/src/test/scala/tailcall/runtime/ConfigPropertySpec.scala [hunk 3]
```diff
@@ -13,7 +14,8 @@ object ConfigPropertySpec extends TailcallSpec with GraphQLTestSpec {
       // Read .graphql files from resources
       // Perform a check from Document to Config and back to Document
       test("config to document identity") {
-        checkAll(graphQLSpecGen("graphql")) { spec =>
+        checkAll(graphQLSpecGen[GraphQLSDLSpec]("graphql")) { spec =>
+          println(spec.name)
```
Comment (tusharmath) [inline runtime/src/test/scala/tailcall/runtime/ConfigPropertySpec.scala:18]: whoopsie!

File: runtime/src/test/scala/tailcall/runtime/ConfigExecutionGraphQLSpec.scala [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import zio.test.Assertion.equalTo
+import zio.test.TestAspect.before
+import zio.test._
+import zio.{Scope, ZIO}
+
+object ConfigExecutionGraphQLSpec extends TailcallSpec with GraphQLTestSpec {
+  override def spec: Spec[TestEnvironment with Scope, Any] = {
+    suite("ExecutionSpec")(test("config to output") {
+      checkAll(graphQLSpecGen[GraphQLExecutionSpec]("graphql")) { spec =>
+        println(spec.name)
```
Comment (tusharmath) [inline runtime/src/test/scala/tailcall/runtime/ConfigExecutionGraphQLSpec.scala:21]: whoopsie

File: runtime/src/main/scala/tailcall/runtime/transcoder/Document2Config.scala [hunk 5]
```diff
@@ -107,7 +107,7 @@ trait Document2Config {
     }
 
   private def toSteps(directives: List[Directive]): List[Operation] = {
-    TValid.foreach(directives)(_.fromDirective[UnsafeSteps]).toOption.flatMap(_.headOption).toList.flatMap(_.steps)
+    directives.flatMap(_.fromDirective[UnsafeSteps].toOption).headOption.map(_.steps).getOrElse(Nil)
```
Comment (sujeetsr) [inline runtime/src/main/scala/tailcall/runtime/transcoder/Document2Config.scala:110]: `validation/test-unsafe-with-http.graphql` fails without this change.

File: runtime/src/main/scala/tailcall/runtime/model/ExpectType.scala [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package tailcall.runtime.model
+
+import tailcall.runtime.DirectiveCodec
+import zio.json.{DeriveJsonCodec, JsonCodec}
+import zio.schema.annotation.caseName
+import zio.schema.{DeriveSchema, Schema}
+
+@caseName("expect")
+final case class ExpectType(output: String)
```
Comment (tusharmath) [inline runtime/src/main/scala/tailcall/runtime/model/ExpectType.scala:9]: If it's json it's going to be easier to write the tests.

File: runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#> server-sdl
+schema {
+  query: Query
+}
+
+type Query {
+  identity: String @http(path: "/resolveWithEnv", query: {data: "{{env.foo}}"}  baseURL: "https://foo.com")
```
Comment (tusharmath) [inline runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql:7]:
```suggestion
  identity: String @http(path: "/resolveWithEnv", query: {data: "{{env.foo}}"}  baseURL: "https://foo.com/echo")
```

Comment (tusharmath) [reply runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql:7]: add a echo route, that returns the the complete request as response. We should be able to assert on method, URL, body, query params etc. using this route.

File: runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+#> client-query
+query @expect(output: """{"identity":"bar"}""") {
```
Comment (tusharmath) [inline runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql:11]:
```suggestion
query @expect(output: """{"method":"GET", "url": "/", "query": {"data": "FOO"}}""") {
```

File: runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#> server-sdl
+schema {
+  query: Query
+}
+
+type Query {
```
Comment (tusharmath) [inline runtime/src/test/resources/tailcall/runtime/graphql/execution/test-resolve-with-env.graphql:6]:
```suggestion
# Option 1
type Request {
  url: String
  method: String
  headers: [Headers]
  body: String
}

# Option 2
scalar Request


type Query {
```

File: runtime/src/test/scala/tailcall/runtime/ConfigValidationSpec.scala [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  def toExpectedError(specValidationError: SpecValidationError): TValid[String, String] = {
+    TValid.fail(specValidationError.message).trace(specValidationError.location: _*)
+  }
+
+  implicit val decoder: JsonDecoder[SpecValidationError] = DeriveJsonDecoder.gen[SpecValidationError]
+}
+
+final case class SpecValidationError(message: String, location: List[String])
```
Comment (tusharmath) [inline runtime/src/test/scala/tailcall/runtime/ConfigValidationSpec.scala:40]: Should support a chunk of errors
```suggestion
final case class SpecValidationError(error: Chunk[Message])
final case class Message(path: Chunk[String], message: String)
```

File: runtime/src/test/resources/tailcall/runtime/graphql/execution/test-dictionary.graphql [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+      value
+    }
+  }
+}
+
+query @expect(output: {z:{a:1}}) {
+  z {
+    a
+  }
+}
```
Comment (arunkmishra) [inline runtime/src/test/resources/tailcall/runtime/graphql/execution/test-dictionary.graphql:54]:
```suggestion
}

```

File: runtime/src/test/resources/tailcall/runtime/graphql/execution/test-inline-field-scalar-type.graphql [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+type Query {
+  foo: String
+}
+
+#> client-query
+query @expect(output: {foo:"Hello"}) {
+  foo
+}
```
Comment (arunkmishra) [inline runtime/src/test/resources/tailcall/runtime/graphql/execution/test-inline-field-scalar-type.graphql:26]:
```suggestion
}

```

File: runtime/src/test/scala/tailcall/runtime/internal/GraphQLTestSpec.scala [hunk 13]
```diff
@@ -1,31 +1,156 @@
...
+case class GraphQLSpec(serverSDL: ValidSDL, client: Option[Client])
+
+object GraphQLSpec {
+
+  case class ValidSDL private (sdl: String)
+
+  object ValidSDL {
+    def fromSDL(sdl: String, fileName: String) = {
+      Parser.parseQuery(sdl).map(_ => new ValidSDL(sdl))
+        .fold(error => throw new Exception(s"${error.toString()}: ${fileName}"), value => value);
```
Comment (arunkmishra) [inline runtime/src/test/scala/tailcall/runtime/internal/GraphQLTestSpec.scala:47]:
```suggestion
        .fold(error => throw new Exception(s"${error.toString()}: ${fileName}"), identity);
```

File: runtime/src/test/scala/tailcall/runtime/internal/GraphQLTestSpec.scala [hunk 14]
```diff
@@ -1,31 +1,156 @@
...
+    val serverSDLZio = ValidSDL.fromSDL(serverSDLStr, file.name)
+
+    val clientSDLZio = ValidSDL.fromSDL(clientSDLStr, file.name)
+
+    val queriesZio =
+      if (!queryString.isBlank()) {
+        ZIO.collectAll {
+          val qs = queryString.split("(?=query)").map(_.trim).toList.map(q => ValidQuery.fromQuery(q, file.name))
+          qs
+        }
```
Comment (arunkmishra) [inline runtime/src/test/scala/tailcall/runtime/internal/GraphQLTestSpec.scala:132]:
```suggestion
        ZIO.collectAll {
          queryString.split("(?=query)").map(_.trim).toList.map(q => ValidQuery.fromQuery(q, file.name))
        }
```

**Pull request closed without merging**